### PR TITLE
feat(packages/jellyfish-api-core): Consortium RPCs

### LINKED ISSUE #1820: Consortium RPCs
Similar to https://github.com/DeFiCh/jellyfish/issues/48 This tracks all the RPC methods and categories being implemented in `jellyfish-api-core`. Please tag update this if you want to work on a feature.

#### What would you like to be added:

...

### PR DESCRIPTION
**Pull request by helloscoopa**
#### What this PR does / why we need it:
/kind feature

#### Which issue(s) does this PR fixes?:

Fixes https://github.com/JellyfishSDK/jellyfish/issues/1820

#### Additional comments?:

- [x] _burntokens_
- [x] _mintokens_
- [x] _setgov_
- [x] _listprices - add tests for pagination_

### CODE REVIEW

**Review by shohamc1 — COMMENTED** [review]
Missing a set of tests that set `mintLimit` and `dailyMintLimit` to `-1`.

https://github.com/DeFiCh/ain/pull/1541/commits/779e30d8dfe7e4e5b0b18f945270d1f3a275725a

File: packages/rich-list-core/__tests__/RichListCore.test.ts [hunk 1]
```diff
@@ -16,8 +16,8 @@ const EXPECTED_RICH_LIST_ADDRESSES = [
   'mud4VMfbBqXNpbt8ur33KHKx8pk3npSq8c',
   'bcrt1qyrfrpadwgw7p5eh3e9h3jmu4kwlz4prx73cqny',
   'bcrt1qyeuu9rvq8a67j86pzvh5897afdmdjpyankp4mu',
-  'mswsMVsyGMj1FzDMbbxw2QW3KvQAv2FKiy',
-  '2NCWAKfEehP3qibkLKYQjXaWMK23k4EDMVS'
+  '2NCWAKfEehP3qibkLKYQjXaWMK23k4EDMVS',
+  'mswsMVsyGMj1FzDMbbxw2QW3KvQAv2FKiy'
```
Comment (cwkang1998) [inline packages/rich-list-core/__tests__/RichListCore.test.ts]: Accidental change?

File: packages/jellyfish-api-core/__tests__/category/oracle/listPrices.test.ts [hunk 2]
```diff
@@ -8,32 +8,23 @@ describe('Oracle', () => {
...
-
-    await container.generate(1)
-  })
-
-  afterAll(async () => {
-    await container.stop()
-  })
-
-  it('should listPrices', async () => {
+  beforeAll(async () => {
```
Comment (cwkang1998) [inline packages/jellyfish-api-core/__tests__/category/oracle/listPrices.test.ts]: Maybe just a minor `nit`, but I think these tests are better for another PR, as they do not directly relate to consortium RPC.

**Review by cwkang1998 — COMMENTED** [review]
Just some minor comments, looks good overall!

File: packages/rich-list-core/__tests__/RichListCore.test.ts (hunk 1 above)
Comment (shohamc1) [reply packages/rich-list-core/__tests__/RichListCore.test.ts]: It's one of the failing tests. This (and the listPrices changes) is already part of https://github.com/JellyfishSDK/jellyfish/pull/1841 which is aiming to resolve compatibility with the v3 release.

File: packages/jellyfish-api-core/__tests__/category/oracle/listPrices.test.ts (hunk 2 above)
Comment (helloscoopa) [reply packages/jellyfish-api-core/__tests__/category/oracle/listPrices.test.ts]: Hmm, yeah I've added it here as just the existing listPrices test was failing... Probably okay having it here I guess..

Comment (cwkang1998) [reply packages/jellyfish-api-core/__tests__/category/oracle/listPrices.test.ts]: Hmm, still think unrelated changes shouldn't be included, but I think these RPC are quite required rn, so let's get it in after its good to go.

**Review by cwkang1998 — COMMENTED** [review]
https://github.com/JellyfishSDK/jellyfish/actions/runs/3478433030/jobs/5822002511#step:5:928
Having some minor test issue here.

**Comment by helloscoopa** [comment]
> https://github.com/JellyfishSDK/jellyfish/actions/runs/3478433030/jobs/5822002511#step:5:928 Having some minor test issue here.

Oh I think this can be ignored right? @shohamc1

**Comment by shohamc1** [comment]
> > https://github.com/JellyfishSDK/jellyfish/actions/runs/3478433030/jobs/5822002511#step:5:928 Having some minor test issue here.
> 
> Oh I think this can be ignored right? @shohamc1

Yup, this is because the image being used is from an RC build. It'll go away once we switch to a release build.

**Review by cwkang1998 — DISMISSED** [review]
:shipit:

File: packages/jellyfish-api-core/src/category/oracle.ts [hunk 3]
```diff
@@ -124,10 +124,10 @@ export class Oracle {
   /**
    * List all aggregated prices.
    *
-   * @param {string} startIndex Optional first key to iterate from, in lexicographical order.
-   * @param {string} includingStart If true, then iterate including starting position. False by default
-   * @param {string} limit Maximum number of orders to return, 100 by default
-   * @return {Promise<ListPricesData[]>} Array of list price data objects
+   * @param {number} startIndex First key to iterate from, in lexicographical order.
+   * @param {boolean} includingStart If true, then iterate including starting position. False by default
+   * @param {number} limit Maximum number of orders to return, 100 by default
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/src/category/oracle.ts]:
```suggestion
   * @param {number} [startIndex] First key to iterate from, in lexicographical order.
   * @param {boolean} [includingStart] If true, then iterate including starting position. False by default
   * @param {number} [limit] Maximum number of orders to return, 100 by default
```

We can do this for optional parameters.

File: packages/jellyfish-api-core/__tests__/category/token/burnTokens.test.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
+import { Testing } from '@defichain/jellyfish-testing'
+import { MasterNodeRegTestContainer } from '@defichain/testcontainers/dist/index'
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/token/burnTokens.test.ts]:
```suggestion
import { MasterNodeRegTestContainer } from '@defichain/testcontainers'
```

File: packages/jellyfish-api-core/__tests__/category/token/burnTokens.test.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    const { txid, vout } = await testing.container.fundAddress(address, 10)
+
+    const burnTxId = await testing.rpc.token.burnTokens(`1@${symbolDBTC}`, address, address, [{ txid, vout }])
+    expect(burnTxId).not.toBe(null)
+
+    await testing.generate(1)
+
+    const tokensAfterBurn = await testing.rpc.account.getAccount(address)
+    expect(tokensAfterBurn[0]).toStrictEqual(`7.00000000@${symbolDBTC}`)
+  })
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/token/burnTokens.test.ts:130]: Are you allowed to burn tokens that you don't own?

You should setup multi masternode testing for this.

File: packages/jellyfish-api-core/__tests__/category/token/mintTokens.test.ts [hunk 6]
```diff
@@ -86,3 +86,355 @@ describe('Token', () => {
...
+
+  it('should throw an error if the token is not specified in governance vars', async () => {
+    // Enable consortium
+    await setGovAttr({ 'v0/params/feature/consortium': 'true' })
+
+    // Set global consortium mint limit for BTC
+    await setGovAttr({
+      [`v0/consortium/${idBTC}/mint_limit`]: '10',
+      [`v0/consortium/${idBTC}/mint_limit_daily`]: '5'
+    })
```
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/token/mintTokens.test.ts]: Your tests are written sequentially; you should avoid that. Each `it('')` should be side-effect free.

This is important to allow tests to be run independently, e.g., 'should throw an error if member daily mint limit exceeds' can't be run standalone; it requires all tests above to be run in sequential order.

```shell
jest --testNamePattern=^Consortium should throw an error if member daily mint limit exceeds --runTestsByPath /@JellyfishSDK/jellyfish/packages/jellyfish-api-core/__tests__/category/token/mintTokens.test.ts
```

**Review by fuxingloh — CHANGES_REQUESTED** [review]

**Review by infinia-yzl — DISMISSED** [review]

**Review by shohamc1 — DISMISSED** [review]

**Review by shohamc1 — DISMISSED** [review]

File: packages/jellyfish-api-core/__tests__/category/token/burnTokens.test.ts [hunk 7]
```diff
@@ -1,16 +1,18 @@
...
-  let account0: string, account1: string
+  let account0: string, account1: string, account2: string
   let idBTC: string
 
   beforeEach(async () => {
     await tGroup.start()
 
     account0 = await tGroup.get(0).generateAddress()
     account1 = await tGroup.get(1).generateAddress()
+    account2 = await tGroup.get(1).generateAddress()
```
Comment (infinia-yzl) [inline packages/jellyfish-api-core/__tests__/category/token/burnTokens.test.ts]: Aren't these the same? @shohamc1

Comment (shohamc1) [reply packages/jellyfish-api-core/__tests__/category/token/burnTokens.test.ts]: Doesn't affect the test, but fixed nonetheless: [f2c2dd1](https://github.com/JellyfishSDK/jellyfish/pull/1860/commits/f2c2dd1c3810a719a0e7a4c96481bd0eb9afe2c9)

**Review by shohamc1 — APPROVED** [review]

**Review by infinia-yzl — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]
Generally the tests LGTM.

**Pull request merged**